test/subscription_map.h

```diff
@@ -0,0 +1,502 @@
+    std::string handle_to_subscription(handle h)  const
+    {
+        std::ostringstream result;
+        bool first_entry = true;
+
+        for(auto const &i: h) {
+            if(!first_entry) {
+                result << "/";
+            }
+
+            result << i.second;
+            first_entry = false;
+        }
+
+        return result.str();
+    }
+
+
+
+};
+
+template<typename Value>
+class single_subscription_map
+        : public subscription_map_base<Value>
+{
+
+public:
+
+    // Handle of an entry
+    using handle = typename subscription_map_base< Value >::handle;
+
+    // Insert a value at the specified subscription path
+    handle insert(MQTT_NS::string_view subscription, Value const &value)
+    {
+        if(!this->find_subscription(subscription).empty()) {
+            throw std::runtime_error(std::string("Subscription already exists in map: ").append(subscription.data(), subscription.size()));
+        }
+
+        auto new_subscription_path = this->create_subscription(subscription);
+        new_subscription_path.back()->second.value = value;
+        return this->path_to_handle(new_subscription_path);
+    }
+
+    // Update a value at the specified subscription path
+    void update(MQTT_NS::string_view subscription, Value const &value)
+    {
+        auto path = this->find_subscription(subscription);
+        if(path.empty()) {
+            throw std::runtime_error("Invalid subscription was specified");
+        }
+
+        path.back()->second.value = value;
+    }
+
+    void update(handle h, Value const &value)
+    {
+        auto entry_iter = this->get_key(h.back());
+        if(entry_iter == this->end()) {
+            throw std::runtime_error("Invalid subscription was specified");
+        }
+        entry_iter->second.value = value;
+    }
+
+    // Remove a value at the specified subscription path
+    void remove(MQTT_NS::string_view subscription)
+    {
+        auto path = this->find_subscription(subscription);
+        if(path.empty()) {
+            return;
+        }
+
+        this->remove_subscription(path);
+    }
+
+    // Remove a value using a handle
+    void remove(handle h)
+    {
+        auto path = this->handle_to_iterators(h);
+        if(!path.empty()) {
+            this->remove_subscription(path);
+        }
+    }
+
+    // Find all subscriptions that match the specified topic
+    template<typename Output>
+    void find(MQTT_NS::string_view topic, Output callback) const
+    {
+        this->find_match(topic, callback);
+    }
+};
+
+
+template<typename Value, class Cont = std::set<Value, std::less<Value>, std::allocator<Value> > >
+class multiple_subscription_map
+    : public subscription_map_base< Cont >
+{
+
+public:
+    // Handle of an entry
+    using handle = typename subscription_map_base< Value >::handle;
+
+    // Insert or update a value at the specified subscription path
+    handle insert_or_update(MQTT_NS::string_view subscription, Value const &value)
```

Comment:
In all code paths inside this function, you are inserting the "value" object.

Shouldn't value be taken by value, and then std::move()'d on the insert function?